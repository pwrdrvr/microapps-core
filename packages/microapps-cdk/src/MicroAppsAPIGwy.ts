import * as apigwy from '@aws-cdk/aws-apigatewayv2';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as iam from '@aws-cdk/aws-iam';
import * as logs from '@aws-cdk/aws-logs';
import * as r53 from '@aws-cdk/aws-route53';
import * as r53targets from '@aws-cdk/aws-route53-targets';
import * as cdk from '@aws-cdk/core';

export interface MicroAppsAPIGwyProps {
  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckes will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: cdk.RemovalPolicy;

  /**
   * CloudFront edge domain name
   *
   * @example apps.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameEdge?: string;

  /**
   * API Gateway origin domain name
   *
   * @example apps-origin.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameOrigin?: string;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * Optional local region ACM certificate to use for API Gateway
   * Note: required when using a custom domain
   *
   * @default none
   */
  readonly certOrigin?: acm.ICertificate;

  /**
   * Route53 zone in which to create optional `domainNameEdge` record
   */
  readonly r53Zone?: r53.IHostedZone;

  /**
   * Path prefix on the root of the API Gateway Stage
   *
   * @example dev/
   * @default none
   */
  readonly rootPathPrefix?: string;
}

export interface IMicroAppsAPIGwy {
  /**
   * Domain Name applied to API Gateway origin
   */
  readonly dnAppsOrigin?: apigwy.IDomainName;

  /**
   * API Gateway
   */
  readonly httpApi: apigwy.HttpApi;
}

export class MicroAppsAPIGwy extends cdk.Construct implements IMicroAppsAPIGwy {
  private _dnAppsOrigin: apigwy.DomainName | undefined;
  public get dnAppsOrigin(): apigwy.IDomainName | undefined {
    return this._dnAppsOrigin;
  }

  private _httpApi: apigwy.HttpApi;
  public get httpApi(): apigwy.HttpApi {
    return this._httpApi;
  }

  /**
   * MicroApps - Create just API Gateway
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: cdk.Construct, id: string, props?: MicroAppsAPIGwyProps) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props cannot be undefined');
    }

    if (
      (props.r53Zone === undefined && props.domainNameEdge !== undefined) ||
      (props.r53Zone !== undefined && props.domainNameEdge === undefined)
    ) {
      throw new Error('If either of r53Zone or domainNameEdge are set then the other must be set');
    }

    if (
      (props.domainNameOrigin === undefined && props.certOrigin !== undefined) ||
      (props.domainNameOrigin !== undefined && props.certOrigin === undefined)
    ) {
      throw new Error(
        'If either of domainNameOrigin or certOrigin are set then the other must be set',
      );
    }

    if (
      (props.domainNameEdge === undefined && props.certOrigin !== undefined) ||
      (props.domainNameEdge !== undefined && props.certOrigin === undefined)
    ) {
      throw new Error(
        'If either of domainNameOrigin or certOrigin are set then the other must be set',
      );
    }

    const {
      r53Zone,
      domainNameEdge,
      domainNameOrigin,
      certOrigin,
      removalPolicy,
      assetNameRoot,
      assetNameSuffix,
      rootPathPrefix,
    } = props;

    // API Gateway uses the `id` string as the gateway name without
    // any randomization... we have to make sure the name is unique-ish
    const apigatewayName = assetNameRoot
      ? `${assetNameRoot}${assetNameSuffix}`
      : `${cdk.Stack.of(this).stackName}-microapps`;

    //
    // APIGateway domain names for CloudFront and origin
    //
    this._httpApi = new apigwy.HttpApi(this, 'gwy', {
      apiName: apigatewayName,
      createDefaultStage: false,
    });
    if (removalPolicy !== undefined) {
      this._httpApi.applyRemovalPolicy(removalPolicy);
    }

    // Create the stage
    const stage = new apigwy.HttpStage(this, 'stage', {
      httpApi: this._httpApi,
      autoDeploy: true,
      // If rootPathPrefix is not defined this will be the $default stage
      stageName: rootPathPrefix,
    });

    if (domainNameEdge !== undefined && certOrigin !== undefined) {
      // Create Custom Domains for API Gateway
      const dnAppsEdge = new apigwy.DomainName(this, 'microapps-apps-edge-dn', {
        domainName: domainNameEdge,
        certificate: certOrigin,
      });
      if (removalPolicy !== undefined) {
        dnAppsEdge.applyRemovalPolicy(removalPolicy);
      }

      // Create the edge domain name mapping for the API
      new apigwy.ApiMapping(this, 'mapping', {
        api: this._httpApi,
        domainName: dnAppsEdge,
        stage,
      });
    }

    if (domainNameOrigin !== undefined && certOrigin !== undefined) {
      this._dnAppsOrigin = new apigwy.DomainName(this, 'origin-dn', {
        domainName: domainNameOrigin,
        certificate: certOrigin,
      });
      if (removalPolicy !== undefined) {
        this._dnAppsOrigin.applyRemovalPolicy(removalPolicy);
      }
    }

    // Enable access logs on API Gateway
    const apiAccessLogs = new logs.LogGroup(this, 'logs', {
      logGroupName: apigatewayName
        ? `/aws/apigwy/${apigatewayName}`
        : `/aws/apigwy/${this.httpApi.httpApiName}`,
      retention: logs.RetentionDays.TWO_WEEKS,
    });
    if (removalPolicy !== undefined) {
      apiAccessLogs.applyRemovalPolicy(removalPolicy);
    }
    // const stage = this._httpApi.defaultStage?.node.defaultChild as apigwy.CfnStage;
    (stage as unknown as apigwy.CfnStage).accessLogSettings = {
      destinationArn: apiAccessLogs.logGroupArn,
      format: JSON.stringify({
        requestId: '$context.requestId',
        userAgent: '$context.identity.userAgent',
        sourceIp: '$context.identity.sourceIp',
        requestTime: '$context.requestTime',
        requestTimeEpoch: '$context.requestTimeEpoch',
        httpMethod: '$context.httpMethod',
        path: '$context.path',
        status: '$context.status',
        protocol: '$context.protocol',
        responseLength: '$context.responseLength',
        domainName: '$context.domainName',
      }),
    };

    // Create a logging role
    // Tips: https://github.com/aws/aws-cdk/issues/11100
    const apiGwyLogRole = new iam.Role(this, 'logs-role', {
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AmazonAPIGatewayPushToCloudWatchLogs',
        ),
      ],
      assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com'),
    });
    apiAccessLogs.grantWrite(apiGwyLogRole);

    //
    // Let API Gateway accept requests using domainNameOrigin
    // That is the origin URI that CloudFront uses for this gateway.
    // The gateway will refuse the traffic if it doesn't have the
    // domain name registered.
    //
    if (this._dnAppsOrigin !== undefined) {
      const mappingAppsApis = new apigwy.ApiMapping(this, 'api-map-origin', {
        api: this._httpApi,
        domainName: this._dnAppsOrigin,
        stage,
      });
      // 2021-12-12 - This should not be needed
      mappingAppsApis.node.addDependency(this._dnAppsOrigin);
      if (removalPolicy !== undefined) {
        mappingAppsApis.applyRemovalPolicy(removalPolicy);
      }
    }

    //
    // Create the origin name for API Gateway
    //
    if (r53Zone !== undefined && this._dnAppsOrigin) {
      const rrAppsOrigin = new r53.ARecord(this, 'origin-arecord', {
        zone: r53Zone,
        recordName: domainNameOrigin,
        target: r53.RecordTarget.fromAlias(
          new r53targets.ApiGatewayv2DomainProperties(
            this._dnAppsOrigin.regionalDomainName,
            this._dnAppsOrigin.regionalHostedZoneId,
          ),
        ),
      });
      if (removalPolicy !== undefined) {
        rrAppsOrigin.applyRemovalPolicy(removalPolicy);
      }
    }
  }
}
