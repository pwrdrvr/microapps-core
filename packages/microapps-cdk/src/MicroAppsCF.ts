import * as apigwy from '@aws-cdk/aws-apigatewayv2';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as cf from '@aws-cdk/aws-cloudfront';
import * as cforigins from '@aws-cdk/aws-cloudfront-origins';
import * as r53 from '@aws-cdk/aws-route53';
import * as r53targets from '@aws-cdk/aws-route53-targets';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { reverseDomain } from './utils/ReverseDomain';

export interface IMicroAppsCF {
  readonly cloudFrontDistro: cf.Distribution;
}

export interface MicroAppsCFProps {
  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckes will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: cdk.RemovalPolicy;

  /**
   * S3 bucket origin for deployed applications
   */
  readonly bucketAppsOrigin: cforigins.S3Origin;

  /**
   * S3 bucket for CloudFront logs
   */
  readonly bucketLogs?: s3.IBucket;

  /**
   * CloudFront Distribution domain name
   *
   * @example apps.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameEdge?: string;

  /**
   * API Gateway custom origin domain name
   *
   * @example apps.pwrdrvr.com
   * @default - retrieved from httpApi, if possible
   */
  readonly domainNameOrigin?: string;

  /**
   * API Gateway v2 HTTP API for apps
   */
  readonly httpApi: apigwy.HttpApi;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * ACM Certificate that covers `domainNameEdge` name
   */
  readonly certEdge?: acm.ICertificate;

  /**
   * Route53 zone in which to create optional `domainNameEdge` record
   */
  readonly r53Zone?: r53.IHostedZone;
}

export class MicroAppsCF extends cdk.Construct implements IMicroAppsCF {
  private _cloudFrontDistro: cf.Distribution;
  public get cloudFrontDistro(): cf.Distribution {
    return this._cloudFrontDistro;
  }

  /**
   * MicroApps - Create just CloudFront resources.
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: cdk.Construct, id: string, props?: MicroAppsCFProps) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props must be set');
    }

    if (
      (props.r53Zone === undefined && props.domainNameEdge !== undefined) ||
      (props.r53Zone !== undefined && props.domainNameEdge === undefined)
    ) {
      throw new Error('If either of r53Zone or domainNameEdge are set then the other must be set');
    }

    const {
      domainNameEdge,
      domainNameOrigin,
      httpApi,
      removalPolicy,
      certEdge,
      assetNameRoot,
      assetNameSuffix,
      r53Zone,
      bucketLogs,
      bucketAppsOrigin,
    } = props;

    //
    // CloudFront Distro
    //
    let httpOriginFQDN: string = 'invalid.pwrdrvr.com';
    let apigwyOriginPolicy: cf.IOriginRequestPolicy = cf.OriginRequestPolicy.ALL_VIEWER;
    if (domainNameOrigin !== undefined) {
      // When using a custom domain we can use the default apigwy origin policy of
      // ALL_VIEWER (all header, all cookies, all query strings)
      httpOriginFQDN = domainNameOrigin;
    } else {
      // When not using a custom domain name we must limit down the origin policy to
      // prevent it from passing the Host header (distribution_id.cloudfront.net) to
      // apigwy which will then reject it with a 403 because it does not match the
      // execute-api name that apigwhy is expecting.
      apigwyOriginPolicy = new cf.OriginRequestPolicy(this, 'apigwy-origin-policy', {
        comment: assetNameRoot ? `${assetNameRoot}-apigwy${assetNameSuffix}` : undefined,
        cookieBehavior: cf.OriginRequestCookieBehavior.all(),
        queryStringBehavior: cf.OriginRequestQueryStringBehavior.all(),
        headerBehavior: cf.OriginRequestHeaderBehavior.allowList('user-agent', 'referer'),
      });
      httpOriginFQDN = `${httpApi.apiId}.execute-api.${cdk.Aws.REGION}.amazonaws.com`;
    }
    const apiGwyOrigin = new cforigins.HttpOrigin(httpOriginFQDN, {
      protocolPolicy: cf.OriginProtocolPolicy.HTTPS_ONLY,
      originSslProtocols: [cf.OriginSslPolicy.TLS_V1_2],
    });
    this._cloudFrontDistro = new cf.Distribution(this, 'cft', {
      comment: assetNameRoot ? `${assetNameRoot}${assetNameSuffix}` : domainNameEdge,
      domainNames: domainNameEdge !== undefined ? [domainNameEdge] : undefined,
      certificate: certEdge,
      httpVersion: cf.HttpVersion.HTTP2,
      defaultBehavior: {
        allowedMethods: cf.AllowedMethods.ALLOW_ALL,
        cachePolicy: cf.CachePolicy.CACHING_DISABLED,
        compress: true,
        originRequestPolicy: apigwyOriginPolicy,
        origin: apiGwyOrigin,
        viewerProtocolPolicy: cf.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      enableIpv6: true,
      priceClass: cf.PriceClass.PRICE_CLASS_100,
      logBucket: bucketLogs,
      logFilePrefix: props.domainNameEdge
        ? `${reverseDomain(props.domainNameEdge)}/cloudfront-raw/`
        : undefined,
    });
    if (removalPolicy !== undefined) {
      this._cloudFrontDistro.applyRemovalPolicy(removalPolicy);
    }

    //
    // Add Behaviors
    //
    const apiGwyBehavior: cf.AddBehaviorOptions = {
      allowedMethods: cf.AllowedMethods.ALLOW_ALL,
      cachePolicy: cf.CachePolicy.CACHING_DISABLED,
      compress: true,
      originRequestPolicy: apigwyOriginPolicy,
      viewerProtocolPolicy: cf.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    };
    const s3Behavior: cf.AddBehaviorOptions = {
      allowedMethods: cf.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
      cachePolicy: cf.CachePolicy.CACHING_OPTIMIZED,
      compress: true,
      originRequestPolicy: cf.OriginRequestPolicy.CORS_S3_ORIGIN,
      viewerProtocolPolicy: cf.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    };
    const apiGwyVersionRootBehavior: cf.AddBehaviorOptions = {
      allowedMethods: cf.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
      cachePolicy: cf.CachePolicy.CACHING_DISABLED,
      compress: true,
      originRequestPolicy: apigwyOriginPolicy,
      viewerProtocolPolicy: cf.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    };

    //
    // Setup Routes
    // Pull to the API first, then pull to S3 if it contains /static/
    // Pull anything under /appName/x.y.z/ folder with '.' in file name to S3
    // Let everything else fall through to the API Gateway
    //
    this._cloudFrontDistro.addBehavior('/*/*/api/*', apiGwyOrigin, apiGwyBehavior);
    this._cloudFrontDistro.addBehavior('/*/*/static/*', bucketAppsOrigin, s3Behavior);
    this._cloudFrontDistro.addBehavior('/*/*/*.*', bucketAppsOrigin, s3Behavior);
    this._cloudFrontDistro.addBehavior('/*/*/', apiGwyOrigin, apiGwyVersionRootBehavior);

    //
    // Create the edge name for the CloudFront distro
    //

    if (r53Zone !== undefined) {
      const rrAppsEdge = new r53.RecordSet(this, 'edge-arecord', {
        recordName: domainNameEdge,
        recordType: r53.RecordType.A,
        target: r53.RecordTarget.fromAlias(new r53targets.CloudFrontTarget(this._cloudFrontDistro)),
        zone: r53Zone,
      });
      if (removalPolicy !== undefined) {
        rrAppsEdge.applyRemovalPolicy(removalPolicy);
      }
    }
  }
}
