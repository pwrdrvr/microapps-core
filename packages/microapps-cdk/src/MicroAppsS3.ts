import { RemovalPolicy } from 'aws-cdk-lib';
import * as cf from 'aws-cdk-lib/aws-cloudfront';
import * as cforigins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

/**
 * Represents a MicroApps S3
 */
export interface IMicroAppsS3 {
  /**
   * S3 bucket for deployed applications
   */
  readonly bucketApps: s3.IBucket;

  /**
   * CloudFront Origin Access Identity for the deployed applications bucket
   */
  readonly bucketAppsOAI: cf.OriginAccessIdentity;

  /**
   * CloudFront Origin for the deployed applications bucket
   */
  readonly bucketAppsOrigin: cforigins.S3Origin;

  /**
   * CloudFront Origin for the deployed applications bucket, used as a signal
   * that this is the primary origin before fallback
   */
  readonly bucketAppsOriginSignal: cforigins.S3Origin;

  /**
   * S3 bucket for staged applications (prior to deploy)
   */
  readonly bucketAppsStaging: s3.IBucket;

  /**
   * S3 bucket for CloudFront logs
   */
  readonly bucketLogs: s3.IBucket;
}

/**
 * Properties to initialize an instance of `MicroAppsS3`.
 */
export interface MicroAppsS3Props {
  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckets will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: RemovalPolicy;

  /**
   * S3 deployed apps bucket name
   *
   * @default auto-assigned
   */
  readonly bucketAppsName?: string;

  /**
   * S3 staging apps bucket name
   *
   * @default auto-assigned
   */
  readonly bucketAppsStagingName?: string;

  /**
   * S3 logs bucket name
   *
   * @default auto-assigned
   */
  readonly bucketLogsName?: string;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * Optional Origin Shield Region
   *
   * This should be the region where the DynamoDB is located so the
   * EdgeToOrigin calls have the lowest latency (~1 ms).
   *
   * @default - none
   */
  readonly originShieldRegion?: string;
}

/**
 * Create the durable MicroApps S3 Buckets
 *
 * These should be created in a stack that will not be deleted if
 * there are breaking changes to MicroApps in the future.
 */
export class MicroAppsS3 extends Construct implements IMicroAppsS3 {
  private _bucketApps: s3.IBucket;
  public get bucketApps(): s3.IBucket {
    return this._bucketApps;
  }

  private _bucketAppsOAI: cf.OriginAccessIdentity;
  public get bucketAppsOAI(): cf.OriginAccessIdentity {
    return this._bucketAppsOAI;
  }

  private _bucketAppsOrigin: cforigins.S3Origin;
  public get bucketAppsOrigin(): cforigins.S3Origin {
    return this._bucketAppsOrigin;
  }

  private _bucketAppsOriginSignal: cforigins.S3Origin;
  public get bucketAppsOriginSignal(): cforigins.S3Origin {
    return this._bucketAppsOriginSignal;
  }

  private _bucketAppsStaging: s3.IBucket;
  public get bucketAppsStaging(): s3.IBucket {
    return this._bucketAppsStaging;
  }

  private _bucketLogs: s3.IBucket;
  public get bucketLogs(): s3.IBucket {
    return this._bucketLogs;
  }

  constructor(scope: Construct, id: string, props?: MicroAppsS3Props) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props must be set');
    }

    const { removalPolicy, assetNameRoot, assetNameSuffix, originShieldRegion } = props;

    // Use Auto-Delete S3Bucket if removal policy is DESTROY
    const s3AutoDeleteItems = removalPolicy === RemovalPolicy.DESTROY;

    //
    // S3 Bucket for Logging - Usable by many stacks
    //
    this._bucketLogs = new s3.Bucket(this, 'logs', {
      bucketName: props.bucketLogsName,
      autoDeleteObjects: s3AutoDeleteItems,
      removalPolicy,
    });

    //
    // S3 Buckets for Apps
    //
    this._bucketApps = new s3.Bucket(this, 'apps', {
      bucketName: props.bucketAppsName,
      autoDeleteObjects: s3AutoDeleteItems,
      removalPolicy,
    });
    this._bucketAppsStaging = new s3.Bucket(this, 'staging', {
      bucketName: props.bucketAppsStagingName,
      autoDeleteObjects: s3AutoDeleteItems,
      removalPolicy,
    });

    // Create S3 Origin Identity
    this._bucketAppsOAI = new cf.OriginAccessIdentity(this, 'oai', {
      comment: assetNameRoot !== undefined ? `${assetNameRoot}${assetNameSuffix}` : undefined,
    });
    if (removalPolicy !== undefined) {
      this._bucketAppsOAI.applyRemovalPolicy(removalPolicy);
    }

    // Add Origin for CloudFront
    this._bucketAppsOrigin = new cforigins.S3Origin(this._bucketApps, {
      originAccessIdentity: this.bucketAppsOAI,
      originShieldRegion,
    });

    this._bucketAppsOriginSignal = new cforigins.S3Origin(this._bucketApps, {
      originShieldRegion,
    });
  }
}
