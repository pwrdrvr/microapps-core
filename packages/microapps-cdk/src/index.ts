export * from './MicroApps';
export * from './MicroAppsAPIGwy';
export * from './MicroAppsCF';
export * from './MicroAppsEdgeToOrigin';
export * from './MicroAppsS3';
export * from './MicroAppsSvcs';
export * from './MicroAppsTable';
