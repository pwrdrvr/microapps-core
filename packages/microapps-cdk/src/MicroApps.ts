import * as acm from '@aws-cdk/aws-certificatemanager';
import * as r53 from '@aws-cdk/aws-route53';
import * as cdk from '@aws-cdk/core';
import { IMicroAppsAPIGwy, MicroAppsAPIGwy } from './MicroAppsAPIGwy';
import { IMicroAppsCF, MicroAppsCF } from './MicroAppsCF';
import { IMicroAppsS3, MicroAppsS3 } from './MicroAppsS3';
import { IMicroAppsSvcs, MicroAppsSvcs } from './MicroAppsSvcs';
import { reverseDomain } from './utils/ReverseDomain';

/**
 * Props for MicroApps
 *
 */
export interface MicroAppsProps {
  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckes will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: cdk.RemovalPolicy;

  /**
   * Passed to NODE_ENV of Router and Deployer Lambda functions.
   *
   * @default dev
   */
  readonly appEnv: string;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * Route53 zone in which to create optional `domainNameEdge` record
   */
  readonly r53Zone?: r53.IHostedZone;

  /**
   * Certificate in US-East-1 for the CloudFront distribution.
   *
   */
  readonly certEdge?: acm.ICertificate;

  /**
   * Certificate in deployed region for the API Gateway.
   *
   */
  readonly certOrigin?: acm.ICertificate;

  /**
   * Use a strict S3 Bucket Policy that prevents applications
   * from reading/writing/modifying/deleting files in the S3 Bucket
   * outside of the path that is specific to their app/version.
   *
   * This setting should be used when applications are less than
   * fully trusted.
   *
   * @default false
   */
  readonly s3StrictBucketPolicy?: boolean;

  /**
   * Applies when using s3StrictBucketPolicy = true
   *
   * IAM Role or IAM User names to exclude from the DENY rules on the S3 Bucket Policy.
   *
   * Roles that are Assumed must instead have their AROA added to `s3PolicyBypassAROAs`.
   *
   * Typically any admin roles / users that need to view or manage the S3 Bucket
   * would be added to this list.
   *
   * @example ['arn:aws:iam::1234567890123:role/AdminAccess', 'arn:aws:iam::1234567890123:user/MyAdminUser']
   *
   * @see s3PolicyBypassAROAs
   */
  readonly s3PolicyBypassPrincipalARNs?: string[];

  /**
   * Applies when using s3StrictBucketPolicy = true
   *
   * AROAs of the IAM Role to exclude from the DENY rules on the S3 Bucket Policy.
   * This allows sessions that assume the IAM Role to be excluded from the
   * DENY rules on the S3 Bucket Policy.
   *
   * Typically any admin roles / users that need to view or manage the S3 Bucket
   * would be added to this list.
   *
   * Roles / users that are used directly, not assumed, can be added to `s3PolicyBypassRoleNames` instead.
   *
   * Note: This AROA must be specified to prevent this policy from locking
   * out non-root sessions that have assumed the admin role.
   *
   * The notPrincipals will only match the role name exactly and will not match
   * any session that has assumed the role since notPrincipals does not allow
   * wildcard matches and does not do wildcard matches implicitly either.
   *
   * The AROA must be used because there are only 3 Principal variables available:
   *  https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html#principaltable
   *  aws:username, aws:userid, aws:PrincipalTag
   *
   * For an assumed role, aws:username is blank, aws:userid is:
   *  [unique id AKA AROA for Role]:[session name]
   *
   * Table of unique ID prefixes such as AROA:
   *  https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_identifiers.html#identifiers-prefixes
   *
   * The name of the role is simply not available for an assumed role and, if it was,
   * a complicated comparison would be requierd to prevent exclusion
   * of applying the Deny Rule to roles from other accounts.
   *
   * To get the AROA with the AWS CLI:
   *   aws iam get-role --role-name ROLE-NAME
   *   aws iam get-user -â€“user-name USER-NAME
   *
   * @example [ 'AROA1234567890123' ]
   *
   * @see s3StrictBucketPolicy
   */
  readonly s3PolicyBypassAROAs?: string[];

  /**
   * Optional custom domain name for the CloudFront distribution.
   *
   * @example apps.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameEdge?: string;

  /**
   * Optional custom domain name for the API Gateway HTTPv2 API.
   *
   * @example apps-origin.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameOrigin?: string;

  /**
   * Path prefix on the root of the CloudFront distribution
   *
   * @example dev/
   */
  readonly rootPathPrefix?: string;

  /**
   * Create an extra Behavior (Route) for /api/ that allows
   * API routes to have a period in them.
   *
   * When false API routes with a period in the path will get routed to S3.
   *
   * When true API routes that contain /api/ in the path will get routed to API Gateway
   * even if they have a period in the path.
   *
   * @default true
   */
  readonly createAPIPathRoute?: boolean;
}

export interface IMicroApps {
  readonly cf: IMicroAppsCF;
  readonly s3: IMicroAppsS3;
  readonly svcs: IMicroAppsSvcs;
  readonly apigwy: IMicroAppsAPIGwy;
}

/**
 * Application deployment and runtime environment.
 */
export class MicroApps extends cdk.Construct implements IMicroApps {
  private _cf: MicroAppsCF;
  public get cf(): IMicroAppsCF {
    return this._cf;
  }

  private _s3: MicroAppsS3;
  public get s3(): IMicroAppsS3 {
    return this._s3;
  }

  private _apigwy: MicroAppsAPIGwy;
  public get apigwy(): IMicroAppsAPIGwy {
    return this._apigwy;
  }

  private _svcs: MicroAppsSvcs;
  public get svcs(): IMicroAppsSvcs {
    return this._svcs;
  }

  /**
   * MicroApps - Create entire stack of CloudFront, S3, API Gateway, and Lambda Functions.
   * This is the "Easy Button" construct to get started as quickly as possible.
   *
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: cdk.Construct, id: string, props?: MicroAppsProps) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props must be set');
    }

    const {
      domainNameEdge,
      domainNameOrigin,
      assetNameRoot,
      assetNameSuffix,
      r53Zone,
      certEdge,
      appEnv = 'dev',
      certOrigin,
      removalPolicy,
      s3PolicyBypassAROAs,
      s3PolicyBypassPrincipalARNs,
      s3StrictBucketPolicy,
      rootPathPrefix,
      createAPIPathRoute = true,
    } = props;

    this._s3 = new MicroAppsS3(this, 's3', {
      removalPolicy,
      bucketLogsName: domainNameEdge ? `${reverseDomain(domainNameEdge)}-logs` : undefined,
      bucketAppsName: domainNameEdge ? `${reverseDomain(domainNameEdge)}` : undefined,
      bucketAppsStagingName: domainNameEdge
        ? `${reverseDomain(domainNameEdge)}-staging`
        : undefined,
      assetNameRoot,
      assetNameSuffix,
    });
    this._apigwy = new MicroAppsAPIGwy(this, 'api', {
      removalPolicy,
      assetNameRoot,
      assetNameSuffix,
      domainNameEdge,
      domainNameOrigin,
      r53Zone,
      certOrigin,
      rootPathPrefix,
    });
    this._cf = new MicroAppsCF(this, 'cft', {
      removalPolicy,
      assetNameRoot,
      assetNameSuffix,
      domainNameEdge,
      domainNameOrigin,
      httpApi: this._apigwy.httpApi,
      r53Zone,
      certEdge,
      bucketAppsOrigin: this._s3.bucketAppsOrigin,
      bucketLogs: this._s3.bucketLogs,
      rootPathPrefix,
      createAPIPathRoute,
    });
    this._svcs = new MicroAppsSvcs(this, 'svcs', {
      httpApi: this.apigwy.httpApi,
      removalPolicy,
      bucketApps: this._s3.bucketApps,
      bucketAppsOAI: this._s3.bucketAppsOAI,
      bucketAppsStaging: this._s3.bucketAppsStaging,
      assetNameRoot,
      assetNameSuffix,
      appEnv,
      s3PolicyBypassAROAs,
      s3PolicyBypassPrincipalARNs,
      s3StrictBucketPolicy,
      rootPathPrefix,
    });
  }
}
