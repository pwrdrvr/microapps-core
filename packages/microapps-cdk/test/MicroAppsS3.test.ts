/// <reference types="jest" />
import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MicroAppsS3 } from '../src/MicroAppsS3';

describe('MicroAppsS3', () => {
  const app = new App({});

  it('works with no params', () => {
    const stack = new Stack(app, 'stack');
    const construct = new MicroAppsS3(stack, 'construct', {});

    expect(construct).toBeDefined();
    expect(construct.bucketApps).toBeDefined();
    expect(construct.bucketAppsStaging).toBeDefined();
    expect(construct.bucketLogs).toBeDefined();
    expect(construct.bucketAppsOAI).toBeDefined();
    expect(construct.bucketAppsOrigin).toBeDefined();

    expect(construct.node).toBeDefined();
    Template.fromStack(stack).resourceCountIs('AWS::S3::Bucket', 3);
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });
});
