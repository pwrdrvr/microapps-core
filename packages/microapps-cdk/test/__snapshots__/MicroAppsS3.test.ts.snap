// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MicroAppsS3 works with no params 1`] = `
Object {
  "Resources": Object {
    "constructappsB8541FB8": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "constructlogsBC77618C": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "constructoai26A46529": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "constructstaging6988EDA5": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
